### PR TITLE
[Parse] Improve recovery from and diagnostics for invalid operator names

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
More accurate diagnostics for invalid or ambiguous characters in operator names that also improve recovery as a consequence. 

Here are some affected examples:
```swift
infix operator =
infix operator ->
infix operator ?
postfix operator !==
infix operator -oo-
infix operator =#=
infix operator ?#$
```

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@rintaro

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -7401,24 +7401,50 @@ Parser::parseDeclOperator(ParseDeclOptions Flags, DeclAttributes &Attributes) {
...
+      // If there's a deprecated body, skip it to improve recovery.
+      if (peekToken().is(tok::l_brace)) {
+        consumeToken();
+        skipSingle();
+      }
+      return nullptr;
+    }
+  // We will handle these either below or in Sema.
+  } else if (Tok.isNot(tok::exclaim_postfix, tok::question_infix,
+                       tok::question_postfix)) {
```
Comment (rintaro) [inline lib/Parse/ParseDecl.cpp]: I think we should handle them as the same as `isAnyOperator()` case. Consider:
```swift
infix operator?$$
```
In this case, the token is `tok:: question_infix`, but we still should diagnose this as `'$$' is considered an identifier ...` right?

**Review by rintaro — CHANGES_REQUESTED** [review]
Nice improvements!

File: lib/Parse/ParseDecl.cpp (hunk 1 above)
Comment (AnthonyLatsis) [reply lib/Parse/ParseDecl.cpp]: Good call!
> I think we should handle them as the same as isAnyOperator() case

Together with `tok::equal` and `tok::arrow`, now that I think about it.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by AnthonyLatsis** [comment]
Thanks for the help in improving the changes!

**Review by rintaro — APPROVED** [review]
Looks good. Thank you Anthony!

**Pull request merged**